FlashApp.sh

```diff
@@ -0,0 +1,247 @@
+    tar -xzf $cloudFWFile -C $WICIMAGEFILE && sync
+    cloudFWFile=$(ls $WICIMAGEFILE/*.wic)
+    if [ -z "$cloudFWFile" ]; then
+        logger "Extracted firmware image file not found; cannot proceed, exiting."
+        exit 1
+    fi
+    # Remove the uncompressed file for space saving.
+    logger "Removing the compressed firmware image file '$compressedFile'"
+    rm -rf $compressedFile && sync
+fi
+
+# Check if the cloudFWFile exists and then print its size
+if [ -f $cloudFWFile ]; then
+    fileSize=$(stat -c %s $cloudFWFile)
+    logger "File '$cloudFWFile' exists and its size is '$fileSize' bytes."
+else
+    logger "File '$cloudFWFile' does not exist; cannot proceed, exiting."
+    exit 1
+fi
+
+logger "Creating the directory '$EXTBLOCK'"
+mkdir -p $EXTBLOCK
+
+# Check if the /boot partition exists; if not, exit.
+boot_partition=$(df | grep "/boot" | cut -d' ' -f1)
+ota_boot_mount_point=$EXTBLOCK/ota_boot
+ota_rootfs_mount_point=$EXTBLOCK/ota_rootfs
+target_rootfs_mount_point=$EXTBLOCK/target_rootfs
+old_boot_bkup=$EXTBLOCK/old_boot_bkup
+
+if [ -z "$boot_partition" ]; then
+    logger "No /boot partition found; cannot proceed, exiting."
+    exit 1
+fi
+
+# check if $old_boot_bkup exists, delete it and create a new one
+if [ -d $old_boot_bkup ]; then
+    logger "Cleaning up the previous /boot partition back-up '$old_boot_bkup'"
+    rm -rf $old_boot_bkup && sync
+fi
+
+mkdir -p $EXTBLOCK/{ota_boot,ota_rootfs,target_rootfs,old_boot_bkup}
+if [ ! -d $ota_boot_mount_point ] || [ ! -d $old_boot_bkup ] || [ ! -d $ota_rootfs_mount_point ] || [ ! -d $target_rootfs_mount_point ]; then
+    logger "Failed to create '$ota_boot_mount_point', '$ota_rootfs_mount_point', '$target_rootfs_mount_point' or '$old_boot_bkup'; cannot proceed."
+    exit 1
+fi
+
+logger "boot_partition: $boot_partition"
+logger "ota_boot_mount_point: $ota_boot_mount_point"
+logger "old_boot_bkup: $old_boot_bkup"
+logger "ota_rootfs_mount_point: $ota_rootfs_mount_point"
+logger "target_rootfs_mount_point: $target_rootfs_mount_point"
+
+# TODO: block the signals to avoid any interruptions during the firmware update.
+
+# back-up the contents of /boot partition
+logger "Backing up the contents of '/boot' partition to '$old_boot_bkup'"
+cp -ar /boot/* $old_boot_bkup/ && sync
+if [ $? -ne 0 ]; then
+    logger "Failed to back-up the contents of '$ota_boot_mount_point' partition; cannot proceed, exiting."
+    exit 1
+else
+    logger "The '/boot/' back-up to '$old_boot_bkup' is successful."
+fi
+
+isRootFSUpdateSuccess=0
+# Mount the WIC file using losetup and get the node name
+losetupNode=$(losetup --find --show --partscan $cloudFWFile)
+if [ -z "$losetupNode" ]; then
+    logger "Failed to setup loop device for $cloudFWFile; cannot proceed, exiting."
+    exit 1
+fi
+logger "Loop device '$losetupNode' is setup for $cloudFWFile"
+
+# mount the P2 partition which is RootFS as readonly to avoid any accidental writes
+ota_rootfs_node=$losetupNode"p2"
```

Comment:
Will it be always p2 partition?

Reply:
Yes, partition is added by [WKS](https://github.com/rdkcentral/meta-vendor-raspberrypi-release/blob/b8430bd92695df1d88065bc73bfceee7c4afb3bd/wic/sdimage-raspberrypi.wks#L5) and it will be creating that in the same order.